RavenDB-23085 Properly track MemoryMapped as we map using the PAL

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23085 
https://issues.hibernatingrhinos.com/issue/RavenDB-23176

### Additional description

Restore tracking of the amount of mapped memory

ayende committed:
RavenDB-23085 Properly track MemoryMapped as we map using the PAL

ayende committed:
RavenDB-23085 Locked memory tracking was moved to the C PAL code

ayende committed:
RavenDB-23085 Making the PAL build update the Directory.Packages.props file

ayende committed:
RavenDB-23085
 - Expose memory lock callback to managed code
 - Only lock once if we have WRITEABLE_MAP
 - Keep the file name around in the handle for reporting purposes in Windows as well
 - Fixing computation of _GetNearestFileSize
 - Properly error on Windows if we cannot lock the memory

arekpalinski commented on src/Voron/Impl/Paging/Pager.cs:
```diff
@@ -332,6 +334,7 @@ public void EnsureContinuous(ref State state, long requestedPageNumber, int numb
         }
         Debug.Assert(totalAllocatedSize >= state.TotalAllocatedSize, "totalAllocatedSize >= state.TotalAllocatedSize");
         state = new State(this, readAddress, writeAddress, totalAllocatedSize, handle);
+        NativeMemory.RegisterFileMapping(state.Pager.FileName, new IntPtr(state.ReadAddress), state.TotalAllocatedSize, null);
```
I wonder whether we shouldn't do it inside `State` constructor. Then both calls:
- `NativeMemory.RegisterFileMapping()`
- `NativeMemory.UnregisterFileMapping()`

would be in the same file. Which actually makes sense since register / unregister should always be called on creation / disposal of a pager state instance.

ayende replied:
Good idea

ayende committed:
RavenDB-23085
 - Moved handling of memory locking failure from native code to managed code
 - Avoid double counting of memory locks on WRITEABLE_MAP files

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085 - Move registration of file mapping to the Pager.State ctor

arekpalinski commented on Directory.Packages.props:
```diff
@@ -84,7 +84,7 @@
     <PackageVersion Include="RabbitMQ.Client" Version="6.8.1" />
     <PackageVersion Include="Raven.CodeAnalysis" Version="1.0.11" />
-    <PackageVersion Include="Raven.Pal" Version="6.2.22" />
+    <PackageVersion Include="Raven.Pal" Version="6.2.41" />
```
Shouldn't it be 7.0.x?

ayende replied:
Yes, but we'll probably want to run that through the numbers - and se it to 7.0 only when we are ready to release.
Because each change consume a cache entry in the NuGet cache, which we cannot easily reuse.

arekpalinski replied:
Makes sense

arekpalinski commented on src/Raven.Pal/Raven.Pal.vcxproj:
```diff
@@ -46,7 +46,7 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="Configuration">
     <ConfigurationType>Application</ConfigurationType>
-    <UseDebugLibraries>false</UseDebugLibraries>
+    <UseDebugLibraries>true</UseDebugLibraries>
```
So in `Release|x64` we use debug libraries? (not sure what does exactly mean though)

ayende replied:
That is used _only_ when I debug this - the actual building process is not related to those values. 
Can safely ignore this.

arekpalinski commented on src/Sparrow.Server/Platform/MemoryLockUsage.cs:
```diff
@@ -0,0 +1,79 @@
+        public static bool RecoverLockedMemoryFailure(Int64 sizeToLock, char* filenamePtr)
+        {
+            if (PlatformDetails.RunningOnPosix)
+                return false; // nothing to do here
```
Can you explain why there is nothing do to on Posix?

ayende replied:
There is no error recovery from locked memory in Linux.
In Windows, we can try to increase the working set size and retry, but on Linux, if it fails, that is it.

arekpalinski commented on test/SlowTests/Issues/RavenDB-23085.cs:
```diff
@@ -0,0 +1,42 @@
+            Assert.True(prev < locked);
+            pager.Dispose();
+            state.Dispose();
+            Assert.Equal(0, locked);
```
Wouldn't this be introducing false positive failures? `MemoryLockUsage.MemoryLockedCalled` can be called by someone else at the time of test execution. I don't think we're guaranteed that it will be always `0`

ayende replied:
Correct, that is a good point. Fixed.

arekpalinski reviewed (commented):
Please check tests failures:

- on Ubuntu - https://github.com/ravendb/ravendb/actions/runs/11953381696/job/33321137870?pr=19633

   
 - an macos  - https://github.com/ravendb/ravendb/actions/runs/11953381696/job/33321138586?pr=19633

arekpalinski commented:
@maciejaszyk can you please take a look at C code?

ayende committed:
RavenDB-23176 Avoid race condition where we may try to truncate the data file if we are concurrently flushing.
Using CreateFileMapping to ensure that we are always just increase the file size

ayende committed:
RavenDB-23176 - Updating binaries

ayende committed:
RavenDB-23085 - Ensure that the test will only account for _our_ changes, not anything else that may happen

ayende committed:
RavenDB-23085 Reversing polarity of check - mlock returns zero on _success_

ayende committed:
RavenDB-23085 Updating binaries

arekpalinski commented:
test this in debug please

arekpalinski commented:
@ayende can you please check failing test

https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/177/testReport/junit/SlowTests.Voron.PalTest/PalJournalTests/TruncateJournal_WhenCalled_ShouldTruncate/

arekpalinski commented:
In addition to a failing test, the issue with `MemoryMapped` being 0 isn't fixed yet (at least on Linux):

https://issues.hibernatingrhinos.com/issue/RavenDB-23085/Memory-info-EncryptionLockedMemory-and-MemoryMapped-is-always-0#focus=Comments-67-1057836.0-0

ayende committed:
RavenDB-23085 - Reworking how we handle writing to the data pager file
We now support write modes (file I/O, mmap, io ring) to handle flushing data into the data pager

- Added locks around pager operations (creation / disposal and writing)
- We now pass the entire set of pages to be written directly to native code for optimal behavior
- Pager's state was split at the native level to global and handle specific state - with ref counting

ayende committed:
RavenDB-23085 - Added support for IO Ring, vectored file IO on linux, as well as mmap & file i/o in both linux & mac

ayende committed:
RavenDB-23085
- Allow to debug PAL using native debugger in Linux
- Fixing invalid flags & handling of IO_Uring setup

ayende committed:
RavenDB-23085 - Updating binaries - using 6.9.1 model to separate from the 6.2 range - will change to 7.0 on release

ayende committed:
RavenDB-23085 - Ensuring we properly reference the static liburing

ayende committed:
RavenDB-23085
- Only use IO_RING if we don't pass writable_map (so we'll actually write using this method)
- Better error handling if we can't lock memory for the ring on Linux
- Simplified selecting the appropriate write mode

ayende committed:
RavenDB-23085 Fixing concurrency issue in test

ayende committed:
RavenDB-23085 - Fixing an issue where writing more than 256 pages at once caused us to skip some

ayende committed:
RavenDB-23085 - Update binaries

ayende committed:
RavenDB-23085 - Better recovery command in the error message

redknightlois commented on src/Raven.Pal/src/posix/pager.c:
```diff
@@ -454,4 +537,54 @@ int32_t rvn_map_memory(void *handle,
+        int64_t offset = buffers[i].page_num * VORON_PAGE_SIZE;
+        int64_t size = (int64_t)buffers[i].count_of_pages * VORON_PAGE_SIZE;
+        if(rvn_pwrite(handle_ptr->file_fd, buffers[i].ptr, size, offset) <= 0)
+        {
```
this is `unlikely` to be a branch taken. I would invert the condition and just `continue`

ayende replied:
Done

redknightlois commented on src/Raven.Pal/src/posix/linuxonly.c:
```diff
@@ -108,5 +114,179 @@ rvn_test_storage_durability(
+        return false;
+
+  int curr_major = 0, curr_minor = 0, curr_patch = 0;
+  sscanf(buffer.release, "%d.%d.%d", &curr_major, &curr_minor, &curr_patch); 
```
In Windows we dont need to use any parsing, this information is available on a bitmap. It seems a bit of a mismeanor to have to do parsing on every pager init. While it is not a soooo common operation, we could just store the value in the process and reuse it.

ayende replied:
Done

redknightlois commented on src/Raven.Pal/src/win/pager.c:
```diff
@@ -119,13 +117,17 @@ int32_t rvn_lock_memory(int32_t open_flags, void *mem, int64_t size, int32_t *de
+            // if we don't consider this as catastrophic error, because otherwise accounting 
+            // for its removal is really complicated
+            g_locked_memory_callback(size, handle->global_state->file_path);
             rc = SUCCESS;
```
Can you explain a bit more what this mean?

ayende replied:
Basically, you may run in a mode that doesn't allow to lock memory
In that case, even though the OS didn't lock the memory, our own accounting would.
This is because it would be much more complex for us to keep track of "locked memory in theory, but not really" and not for accounting purposes.

Note that this usually failing to lock memory is a fatal error for us, and you need to configure this especially.

redknightlois replied:
All that context should be in the code. I think it is important to know.

ayende replied:
It's there, that is what the comment is saying :-)

redknightlois replied:
I mean the context of this... Now I understand what that comment means :D

redknightlois commented on src/Raven.Pal/src/win/pager.c:
```diff
@@ -558,3 +696,176 @@ int32_t rvn_pager_get_file_handle(
+            if (FAILED(hr))
+            {
+                *detailed_error_code = hr;
+                return FAIL_IO_RING_WRITE;
```
I am not that well versed into `EnterCriticalSection` but it seems that if write fails, we are not going out from it.

ayende replied:
You are 100% correct, yes.

ayende committed:
RavenDB-23085
- Fixed windows truncate file handling
- Code gen improvement for likely branch

ayende committed:
RavenDB-23085
- Fixing IO error not releasing the lock
- Added locking for IO Ring writes in Linux
- Cached the check for kernel version for io ring

ayende committed:
RavenDB-23085 - Changing how we call IO Ring calls in Windows to use dynamic load, to avoid faulting on older versions

ayende committed:
RavenDB-23085 - Updating binaries

garayx commented:
test this please

ayende committed:
RavenDB-23085 - Allowing to build PAL using MSVC, not just zig cc

ayende committed:
RavenDB-23085 updating binaries

ayende committed:
RavenDB-23085
- Started to implement shared jounrals
- We can now submit a journal write from a branch env to a root env, which will record it in its own journal

ayende committed:
RavenDB-23085 - Simplifying how we track the last flushed state for syncing to the headers files

ayende committed:
RavenDB-23085
- Changing the way we write journals in PAL / native code so we can accept multiple entries to the journal at once
- Ensure that initial setup of the IoRing functions in Windows is thread safe
- Adding a way to generate hard links in PAL

ayende committed:
RavenDB-23085
- Better name for the system store when running in memory
- Updating journal writes to the new API
- Better handling of submission of journal entry to be written by root env
- Separated how we detect if transaction requires writing to the journal
- An empty transaction may still want to write to the journal, if there are pending writes from branch env.
- Adding test showcasing merge journals

ayende committed:
RavenDB-23085
- Updating binaries

ayende committed:
RavenDB-23085
- Implement shared journals queue with a proper error handling / waiting
- Will hard-link files during shared journals to reduce the number of overall journals in use

ayende committed:
RavenDB-23085
- Fixing tests to match actual behavior

ayende committed:
RavenDB-23085
- Adding DatabaseId to TransactionHeader and FileHeader
- Updating Voron schema version
- Refactoring how we read the headers and verify them

ayende committed:
RavenDB-23085
- Added DbId to the header.one and header.two files (in a forward compatible manner) so we can use the db id during recovery
- Reading of the headers files is now done & validated in a single location, and greatly simplified
- Can now register in the TransactionHeader which db it belongs to
-
- Finished refactoring how we deal with header accessor without pointers
- Copy db id to the header when compacting storage

ayende committed:
RavenDB-23085 - Removed journal recycling feature - because we want to use shared journals, and those two are in conflict with one another

ayende committed:
RavenDB-23085
- Updated journal recovery code to take into account shared journals

ayende committed:
RavenDB-23085
- Do not try to open a journal writer if you are not the root journal
- Fixing test assertion

ayende committed:
RavenDB-23085 - To handle recovery and writing to the same journal file after restart, we now check if the file is already a hard link, and the wire it the same journal entry in the branch in that case

ayende committed:
RavenDB-23085 - updating binaries

ayende committed:
RavenDB-23085 - Ensuring that we can upgrade from a legacy database to the new format seamlessly

ayende committed:
RavenDB-23085
- When we read a legacy journal, we'll not keep using it, but create a _new_ file - so all root environment can immediately share with their branches
- Refactored how we deal with recovery
- Adding a bunch of tests to verify various migration from legacy scenarios

ayende committed:
wip

ayende committed:
RavenDB-23085
- Fixing jounral typo
- Avoid repeated values meaning the same thing in env record
- Move updating the journal position from JournalFile to the WriteAheadJournal so we can properly track across shared journals
- Transaction headers are now properly accounted for on a per journal file instance per environment, and not globally
- JournalFile now has (global across envs) flag that tell us if we are done writing to it, so both branch and root can easily discard completed journals
- When there is no active transaction and we are flushing, we take the chance to discard all completed journals immediately, in both branch & root env.

ayende committed:
RavenDB-23085
- Properly resetting the current file when we are done with it

ayende committed:
RavenDB-23085
- Wiring the indexes as branches to the root of the database - so we have a single shared journal and reduced writes.

ayende committed:
RavenDB-23085 - Reversing polarity of check

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085 - Removing Recyclable Journals remnants

maciejaszyk commented on src/Raven.Pal/src/posix/linuxonly.c:
```diff
@@ -108,5 +114,240 @@ rvn_test_storage_durability(
+            rc = _submit_and_wait(&handle_ptr->global_state->ring, submitted, detailed_error_code);
+            if (rc != SUCCESS)
+            {
+                return rc;
```
What is the expected behavior when we fail? We returned error code and never unlocked the mutex. Is there a path to unlock it later?

ayende replied:
Failure here is the same a fatal error, and will force a restart of the environment.
But *should* release the lock, yes.

ayende committed:
RavenDB-23085
- Properly handle journal sizes that are not multiple of 8KB but 4KB
- Deal with transaction that ends in the last 4KB of the journal

ayende committed:
RavenDB-23085 - Ensuring we properly cleanup on error in PAL

ayende committed:
RavenDB-23085 - Updating binaries

ayende committed:
RavenDB-23085 - Fixing recovery while dealing with corrupted journals

ayende committed:
RavenDB-23085
- Ensuring that we properly mark journals as done when we recover a database
- Removing debug code

ayende committed:
RavenDB-23085
 - Ensuring that we properly scope the shared journals so we avoid trying to use them after the caller exited the scope

ayende committed:
RavenDB-23085
- Changed disposal order of DocumentDatabase so we'll first dispose of the indexes, and only then the transaction merger, so we'll let the indexes flush their changes to the shared journal
- Ensure that we are properly setting the scope for the database transaction merger

ayende committed:
RavenDB-23085 - Move handling of cancellation to the root journal SubmitBranchJournalEntry

ayende committed:
RavenDB-23085
- Separated TxMerger shutdown token from DatabaseShutdown cancellation token - because we want to stop the transaction merger later
- Transaction merger is now aware that it should run transactions if there is a pending journal write
- Cancellation Token for merged commits is now external to the journal

ayende committed:
RavenDB-23085
- Disable recovery & integrity checks when we are in VerifyNoUnexpectedValidTransactionsAfter
- DisableOnRecoveryErrorHandler & DisableOnIntegrityErrorOfAlreadySyncedDataHandler now use a non-null disabling method, so we won't throw if we detect an error while disabled
- Fixed the test corrupting the journal so it would corrupt the journal from that point _forward_

ayende committed:
RavenDB-23085
- Moving test throw hook to just before actual write to file

ayende committed:
RavenDB-23085 - Fixing IncrementalBackup - will now handle reading of the database id in a snapshot scenario properly

ayende committed:
RavenDB-23085
- Changing to using dedicated JournalId per environment that is not exposed at any higher level
- This allows far simpler mechanism with regards to moving databases & indexes, restoring from snapshot backups, etc

ayende committed:
RavenDB-23085
- Moved join of index journal to database journal in root / branch mode to InitializeOptions so will apply to new indexes from the start
- Added a test to check that we are using hard links for indexes from the get go
- When done with shared journal scope - ensure that we replace the cancellation token with an already canceled one
- Properly handle recovering from journal that has transactions only for a branch and not the root
- Removed debug code that no longer makes sense given root/ branch usage

ayende committed:
RavenDB-23085
- Report on debug endpoint whatever an environment is a root or a branch
- Move hard link detection to the StorageEncryptionOptions
- Use a linked cancellation token source to ensure that we are properly canceled when exiting shared journal state.
- Ensuring that the branch will wait until the root releases it, to avoid accessing released memory
- Fixing tests

ayende committed:
RavenDB-23085
- Fixing test reliance on env variable that isn't reuqired.

ayende committed:
RavenDB-23085
- Fixing snapshot restore - we'll force the creation of a _new_ journal file after to restore, to prevent linking the indexes to an old journal file
- Fixing tests

ayende committed:
RavenDB-23085
- Changed how we handle avoiding merging a shared journal that already contains entries from another env that aren't inline (snapshots taken at different times)
- We now detect if the journal file contains the environment at recovery time and if they do, we only allow to use the journal for that environment if a hard link already exists.

ayende committed:
RavenDB-23085
- Added configuration option to disable shared journals
- Disable shared journals for a test that manually tie between transaction merger commands and indexing progress

ayende committed:
RavenDB-23085
- Set the option AvoidSharedJournals globally for the test
